public/app/plugins/datasource/cloudwatch/utils/datalinks.ts

```diff
@@ -0,0 +1,88 @@
+import { DataFrame, DataLink, DataQueryRequest, DataQueryResponse, ScopedVars, TimeRange } from '@grafana/data';
+import { CloudWatchLogsQuery, CloudWatchQuery } from '../types';
+import { AwsUrl, encodeUrl } from '../aws_url';
+import { getDataSourceSrv } from '@grafana/runtime';
+
+type ReplaceFn = (
+  target?: string,
+  scopedVars?: ScopedVars,
+  displayErrorIfIsMultiTemplateVariable?: boolean,
+  fieldName?: string
+) => string;
+
+export async function addDataLinksToLogsResponse(
+  response: DataQueryResponse,
+  request: DataQueryRequest<CloudWatchQuery>,
+  range: TimeRange,
+  replaceFn: ReplaceFn,
+  getRegion: (region: string) => string,
+  tracingDatasourceUid?: string
+): Promise<void> {
+  const replace = (target: string, fieldName?: string) => replaceFn(target, request.scopedVars, true, fieldName);
+
+  for (const dataFrame of response.data as DataFrame[]) {
+    const curTarget = request.targets.find((target) => target.refId === dataFrame.refId) as CloudWatchLogsQuery;
+    const interpolatedRegion = getRegion(replace(curTarget.region, 'region'));
+
+    for (const field of dataFrame.fields) {
+      if (field.name === '@xrayTraceId' && tracingDatasourceUid) {
+        getRegion(replace(curTarget.region, 'region'));
+        const xrayLink = await createInternalXrayLink(tracingDatasourceUid, interpolatedRegion);
+        if (xrayLink) {
+          field.config.links = [xrayLink];
+        }
+      } else {
+        // Right now we add generic link to open the query in xray console to every field so it shows in the logs row
+        // details. Unfortunately this also creates link for all values inside table which look weird.
```

Comment:
This will probably be changed in the future but did not want to change it in this PR